### ISSUE TITLE
CONNECT description in Semantics is a bit 1.1 specific

### ISSUE THREAD

**Issue by quasicomputational** [httpwg/http-core#183]
The description of the CONNECT methods in Semantics (section 7.3.6) is has some HTTP/1.1 specific language in it: "Any 2xx (Successful) response indicates that the sender (and all inbound proxies) will switch to tunnel mode immediately after the blank line that concludes the successful response's header section; data received after that blank line is from the server identified by the request-target."

Relatedly, there's no description of CONNECT in the Messaging draft.

**Comment by wtarreau** [comment]
I think it would be useful to explain it differently, that basically CONNECT method creates a tunnel, that it doesn't carry any payload in either direction, and that they payload is replaced with the bidirectional stream of bytes. As such it behaves as if there was an infinite payload in parallel in both directions. That should help map it better on top of H1 and H2 : for H1 it starts with the first byte following the empty line ; for H2 it implies it's transported over DATA frames.

**Comment by LPardue** [comment]
Agree with @wtarreau.

I think in the HTTP/3 draft there was an issue raised that the tunnel doesn't forward packets but instead exchanges transport PDU payload

**Comment by MikeBishop** [comment]
During our discussion about message boundary delineation, I came to a similar conclusion that CONNECT doesn't happen *after* an HTTP message requesting the tunnel, it's a request with a body of unknown size which MUST be processed in a streaming fashion.

**Comment by wtarreau** [comment]
@MikeBishop you're right, it's not *after* an HTTP message because its content-length must be ignored, which means that the tunnel doesn't start after any supposed payload but really after the headers. From an implementation perspective, I like to see it as infinite payload for request and response because it usually more or less matches the way the frames (or byte streams) have to be handled.

**Comment by awwright** [comment]
> ... for H2 it implies it's transported over DATA frames.

Wouldn't this imply that if I added `Transfer-Encoding: chunked` to a CONNECT request, that the server would have to decode the chunked encoding before getting to raw tunnel data?

**Comment by mnot** [comment]
Waiting for #340.

**Comment by mnot** [comment]
This seems to have been dealt with in other issues; the text in Semantics now looks pretty version-neutral.

Messaging doesn't have a dedicated section about CONNECT, but does mention it in 6.3 Message Body Length; does it need more than that?

**Comment by wtarreau** [comment]
I think we're good as is.

**Comment by mnot** [comment]
Closing; if anyone sees something, please ask for reopening, or file another issue.
